### PR TITLE
Flatten inheritance for DeepSpeed strategy

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

DeepSpeedStrategy now inherits from ParallelStrategy instead of DDPStrategy.

Fixes #[11863](https://github.com/PyTorchLightning/pytorch-lightning/issues/11863)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edward-io** [comment]
Update typing.

**Comment by edward-io** [comment]
Rebase, reorder methods.

File: pytorch_lightning/strategies/deepspeed.py [hunk 1]
```diff
@@ -334,17 +425,39 @@ def _load_config(self, config):
                 config = json.load(f)
         return config
 
-    def setup_distributed(self):
-        reset_seed()
+    def set_world_ranks(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/deepspeed.py:428]: same comment as elsewhere, we can make this protected

File: pytorch_lightning/strategies/deepspeed.py [hunk 2]
```diff
@@ -354,6 +467,26 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
 
+    def barrier(self, *args, **kwargs) -> None:
+        if not distributed_available():
+            return
+        if _TORCH_GREATER_EQUAL_1_8 and torch.distributed.get_backend() == "nccl":
+            torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
+        else:
+            torch.distributed.barrier()
+
+    def setup_distributed(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/deepspeed.py:478]: can make this protected

File: pytorch_lightning/strategies/deepspeed.py [hunk 3]
```diff
@@ -771,16 +838,32 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
...
-            )
-        return False
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        obj = [obj]
+        if self.global_rank != src:
+            obj = [None]
+        torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
+        return obj[0]
+
+    def model_to_device(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/deepspeed.py:848]: super nit: same as elsewhere, let's move this next to `setup`

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/strategies/deepspeed.py (hunk 1 above)
Comment (edward-io) [reply pytorch_lightning/strategies/deepspeed.py:428]: @ananthsub i don't think we can because Accelerator Connector calls it in the _lazy_init_strategy.

Comment (ananthsub) [reply pytorch_lightning/strategies/deepspeed.py:428]: ah I didn't realize that!

**Review by tchaton — COMMENTED** [review]
@edward-io @ananthsub  I am still failing to understand the full reasoning behind this initiative.

Inheritance is a strong mechanism to enforce the right abstraction design and code sync between classes and the fact it might break is a great sign. It means the abstraction isn't just right yet and work is left to be done. 

Now, we are back to square one. If I improve the deadlock mechanism, I need to change it in all DDP-based strategies and am likely to forget one. This happened countless times in the past. 
 
I know the long-term goal is to move everything into plugins and mixin to remove code duplication, but bugs are likely to come in the meantime as this is a long refactor.

Lightning is meant to have the most stable DDP out there, any DDP bugs will damage the PyTorch Lightning brand.

I would prefer for us to abstract the duplicated code first (DDP deadlock reconciliation, collectives, launchers, etc...) before doing the flattening. This seems quite a high-risk initiative to me and done in the wrong order, especially right before the v1.6 release.

IMO, I would prefer to de-scope this for 1.6 to avoid any bad surprises.

@carmocca

File: pytorch_lightning/strategies/deepspeed.py [hunk 4]
```diff
@@ -321,6 +337,83 @@ def __init__(
         self.hysteresis = hysteresis
         self.min_loss_scale = min_loss_scale
 
+    @property
+    def num_nodes(self) -> int:
+        return self._num_nodes
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:345]: Seems to be missing `is_distributed` property. This means DistributedSampler won't be applied.

File: pytorch_lightning/strategies/deepspeed.py [hunk 5]
```diff
@@ -321,6 +337,83 @@ def __init__(
         self.hysteresis = hysteresis
         self.min_loss_scale = min_loss_scale
 
+    @property
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:340]: Let's be consistent and order property as DDP. This would make debugging and readability simpler.

File: pytorch_lightning/strategies/deepspeed.py [hunk 6]
```diff
@@ -334,7 +427,19 @@ def _load_config(self, config):
...
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
+    def _configure_launcher(self) -> None:
+        self._launcher = _SubprocessScriptLauncher(self.cluster_environment, self.num_processes, self.num_nodes)
+        if not self.cluster_environment.creates_processes_externally:
+            self._rank_0_will_call_children_scripts = True
+
+    def setup_environment(self) -> None:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:442]: Let's follow DDP structure with `setup_environment` and `setup_distributed` for consistency.

File: pytorch_lightning/strategies/deepspeed.py [hunk 7]
```diff
@@ -492,24 +627,6 @@ def _initialize_deepspeed_train(self, model):
             self.lr_scheduler_configs = [lr_scheduler]
         self.model = model
 
-    @contextlib.contextmanager
-    def model_sharded_context(self) -> Generator[None, None, None]:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:496]: Why was it moved?

File: pytorch_lightning/strategies/deepspeed.py [hunk 8]
```diff
@@ -354,6 +479,24 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
 
+    def model_to_device(self) -> None:
+        log.detail(f"{self.__class__.__name__}: moving model to device [{self.root_device}]...")
+        self.model.to(self.root_device)
+
+    def barrier(self, *args, **kwargs) -> None:
+        if not distributed_available():
+            return
+
+        if self.root_device.type == "cpu":
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:490]: Why it is different from master? Let's have this change into its own Pr

File: pytorch_lightning/strategies/deepspeed.py [hunk 9]
```diff
@@ -372,10 +515,6 @@ def _set_node_environment_variables(self) -> None:
         os.environ["WORLD_SIZE"] = str(self.world_size)
         os.environ["LOCAL_RANK"] = str(self.local_rank)
 
-    @property
-    def restore_checkpoint_after_setup(self) -> bool:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:376]: Why was it moved?

File: pytorch_lightning/strategies/deepspeed.py [hunk 10]
```diff
@@ -458,10 +597,6 @@ def _init_optimizers(self) -> Tuple[Optimizer, Optional[LRSchedulerConfig], Opti
             optimizer_frequencies[0] if optimizer_frequencies else None,
         )
 
-    @property
-    def zero_stage_3(self) -> bool:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:462]: Why was it moved ?

File: pytorch_lightning/strategies/deepspeed.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/strategies/deepspeed.py:376]: Note: Moving code makes it harder for the reviewer to understand what is added and just moved. I would advise keeping the code re-organization to another PR.

File: pytorch_lightning/strategies/deepspeed.py [hunk 11]
```diff
@@ -771,16 +848,28 @@ def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
             )
         return client_state
 
-    @property
-    def lightning_restore_optimizer(self) -> bool:
```
Comment (tchaton) [inline pytorch_lightning/strategies/deepspeed.py:775]: Same here.

**Comment by four4fish** [comment]
@tchaton The flatten inheritance for Strategy was voted to be one of the top projects in grid&meta roadmap https://www.notion.so/gridai/1f24be0d26e84adaa064ade0c39e50cc?v=4bab7db846714ff98e72b8a9c62b5050 (we don't have permission anymore, you guys still could see this?)

I feel flatten the inheritance is super helpful for stable strategy support and avoiding bugs. Silent overrides or call order (where to call super.xfunction) in the overrides function is always error prone. More important, it's very hard to trace what logic exactly happening for each strategies with this inheritance structure.

For example: 
1. DDP has configure_ddp() function (which supports localSGD) has been inherited by deepspeed and horovod, which could be misleading
2. A while ago, @ananthsub found it’s very not obvious to tell whether we are move_model_to_device first or setup_optimizer first for some strategies, because the inheritance structure and the inconsistent place where each strategy calls their super.xfunction. If all the logic clearly listed in each strategies, we may found [this bug](https://github.com/PyTorchLightning/pytorch-lightning/pull/11886) way earlier

**Comment by Borda** [comment]
@carmocca @awaelchli, how do we this about this one?
